benchmark/result/result.py

```diff
@@ -0,0 +1,29 @@
+"""Basic result classes for benchmarking."""
+
+from uuid import UUID
+
+from attrs import define, field
+from attrs.validators import instance_of
+from pandas import DataFrame
+
+from benchmark.result.utils import _convert_metadata_to_string
+
+
+@define(frozen=True)
+class Result:
+    """A single result of the benchmarking."""
+
+    benchmark_origin_id: UUID = field(validator=instance_of(UUID))
+    """The unique identifier of the benchmark running which can be set
+    to compare different executions of the same benchmark setting."""
+
+    metadata: dict[str, str] = field(
+        validator=instance_of(dict), converter=_convert_metadata_to_string
+    )
+    """Metadata about the benchmarking result."""
+
+    benchmark_result: DataFrame = field()
+    """The result of the benchmarked callable."""
+
+    execution_time_ns: int = field(validator=instance_of(int))
+    """The execution time of the benchmark in nanoseconds."""
```

Comment:
We can also convert it into secounds directly. Nanoseconds is just the unit one gets when calling the performance counter function if you want to avoid float precision problems and I havn't touched it.

Reply:
```suggestion
    execution_time_sec: float = field(validator=instance_of(float))
    """The execution time of the benchmark in seconds."""

```

Reply:
+1 for integers

Reply:
The datatype is not related to my argument

My point is that we will rarely have anything to do with benchmarks that will even be remotely on the order of that short timescale. I dont care if `perf_counter` (float) or `perf_counter_ns` (int) is used

Its kind of measuring the width of a house and storing it in nanometers -> why

Reply:
I think it's very much the same rationale like why our `MemoryInfo` class internally stores stuff in bytes. Also there, stuff will pretty much never be in the scale of bytes but significantly bigger. But to me it makes sense to store stuff in the smallest unit and then use `int`s for it. It's like when people represent monetary values like 314 cents (i.e. int) instead of 3.14 euros (i.e. floats) for the reason fabian mentioned: no rounding problems. Note that this doesn't mean that we have to work with ns though, just as we don't work with bytes in the memory case. For example, we could simply add a property in seconds that does the conversion (or a function that gets the desired format as input) + optionally just hide the ns attribute by making it private. That would separate the representation from the user interface. Thoughts?

Reply:
in principle it is also flawed in the `MemoryInfo` class but there the gap between stored unit and realistic unit is maybe 3 orders of magnitude (kbyte prob not unrealistic for a searchspace), here its at least 9, realisticly 10-11 orders of magniture (being generous and expecting something to run only 10s).

If you the base unit of storage is so far away from the expected unit we need then we could actually also run into floating point problems on the other (higher) end. 

Can you tell me what `rounding problems` you would actually expect? I would expect the least amount of rounding problems exactly when the unit of storage is as close as possible to the expected unit required to represent most results, not by using the smallest unit possible.

Reply:
OK, perhaps one important thing upfront: I think we all agree that this particular situation here is not really critical. So if you are really against `ns`, then I won't even try to convince you and we go with seconds instead.

-----
_That said: "Discussion mode on"_

But just from a conceptual level, I think its – in general – the right thing to do to represent quantities in their original unit. The rounding problem you'll encounter is the typical thing with floats not being able to accurately represent factions, i.e the common `0.1 + 0.2` kind of problem. If you convert to float before storing, that's the point where you will already loose information. Consider
```python
# This is the raw data
time_in_ns: int = 2134012431963453443

# But we store this instead
time_in_s: float = time_in_ns / 1_000_000_000

# We can no longer reconstruct the exact measurements
time_in_ns_reconstructed = int(time_in_s * 1_000_000_000)

print(time_in_ns_reconstructed)
# --> 2134012431963453440
```
So the original number is just lost.

Again: we all agree that in this context the resolution does not matter. But these small differences can cause troubles in other contexts, e.g. equality comparison / serialization etc.

The other rounding concern you had is actually none: ints are open-ended in python, and converting to any other unit on the fly will not cause any additional loss of information as the one from above that you'd get right at storage time.

_"Discussion mode off"_

----

Reply:
I also think that `ns` is not necessary: The idea is to have the benchmarks for stuff that is quite complex, has multiple MC runs and so on. I think a comparison on seconds level and using `int` is fully sufficient.

Reply:
the combination int and perf_counter makes no sense. Let's settle for regular float and perf_counter, measuring in seconds. I think we should be able to all agree with that?